File: src/MixNetFx/FacInterfaces/frmProductionDataConfigureDialogBase.vb
Imports Facesso
Imports Facesso.Data
Imports System.Windows.Forms
Imports System.Drawing

Public Class frmProductionDataConfigureDialogBase

    Private myTaskItem As IFacessoImportTaskItem
    Private myLabourValues As LabourValueInfoCollection
    Protected myAllowMultipleAssignments As Boolean

    Public Function HandleDialog(ByVal TaskItem As IFacessoImportTaskItem) As DialogResult
        myTaskItem = TaskItem
        myLabourValues = LabourValueInfoCollection.GetWorkGroupAssignedLabourValues( _
            FacessoGeneric.LoginInfo.IDSubsidiary, myTaskItem.ForWorkgroup)

        InitializeControls()
        RebuildLists()
        Me.ShowDialog()
        Return Me.DialogResult
    End Function

    Protected Property TaskItem() As IFacessoImportTaskItem
        Get
            Return myTaskItem
        End Get
        Set(ByVal value As IFacessoImportTaskItem)
            myTaskItem = value
        End Set
    End Property

    Protected Overridable Sub InitializeControls()
        With lvwDeviceItems
            .Columns.Add("ID", -2, HorizontalAlignment.Left)
            .Columns.Add("Beschreibung", -2, HorizontalAlignment.Left)
            .Columns.Add("Arbeitswert", -2, HorizontalAlignment.Left)
        End With

        If myTaskItem.ConversionItems Is Nothing Then
            myTaskItem.ConversionItems = myTaskItem.ConversionItemsDelegate.Invoke()
        End If
        lblTitel.Text = "Konfiguration für Produktiv-Site:" & vbNewLine & _
                      myTaskItem.ForWorkgroup.ListItemText
    End Sub

    Protected Overridable Sub RebuildLists()

        Dim locSelectedLabourValue As LabourValueInfo = Nothing
        Dim locSelectedDeviceItem As IFacessoConversionItem = Nothing

        Try
            locSelectedLabourValue = ucLabourValues.FirstSelectedLabourValue
            locSelectedDeviceItem = DirectCast(lvwDeviceItems.SelectedItems(0).Tag, IFacessoConversionItem)
        Catch ex As Exception
            'Nur Fehler abfangen. Falls nichts selektiert war,
            'ist das entsprechende Item eh Nothing.
        End Try

        lvwDeviceItems.BeginUpdate()
        lvwDeviceItems.Items.Clear()
        'Nur neu aufbauen, wenn die ableitende Klasse
        'diese Eigenschaft auf True setzt.
        If Not BlockDeviceListBuilding Then
            If myTaskItem.ConversionItems IsNot Nothing Then
                For Each locItem As IFacessoConversionItem In myTaskItem.ConversionItems
                    Dim locLvwItem As New ListViewItem(locItem.AlienElementID.ToString("000000"))
                    locLvwItem.SubItems.Add(locItem.Itemname)
                    If locItem.HomeElementID = -1 Then
                        locLvwItem.SubItems.Add("- - -")
                    Else
                        locLvwItem.SubItems.Add(locItem.HomeElementName)
                        locLvwItem.Font = New Font(locLvwItem.Font, FontStyle.Bold)
                    End If
                    locLvwItem.Tag = locItem
                    lvwDeviceItems.Items.Add(locLvwItem)
                Next
            End If

            lvwDeviceItems.Columns(0).Width = -2
            lvwDeviceItems.Columns(1).Width = -2
            lvwDeviceItems.Columns(2).Width = -2
        End If
        lvwDeviceItems.EndUpdate()

        Dim locToAssignList As New LabourValueInfoCollection
        For Each locItem As LabourValueInfo In myLabourValues
            locToAssignList.Add(locItem)
        Next

        'Elemente, die schon verwendet wurden, nur dann entfernen, wenn 
        'Mehrfachzuweisungen nicht erlaubt sind!
        If Not myAllowMultipleAssignments Then
            For Each locDestItem As ListViewItem In lvwDeviceItems.Items
                Dim locItem As IFacessoConversionItem = DirectCast(locDestItem.Tag, IFacessoConversionItem)
                locToAssignList.Remove(New ActiveDev.IntKey(locItem.HomeElementID))
            Next
        End If
        ucLabourValues.LabourValues = locToAssignList

        'Dafür sorgen, dass zuvor Dargestellte wieder angezeigt werden.
        If locSelectedLabourValue IsNot Nothing Then
            ucLabourValues.SelectLabourValue(locSelectedLabourValue, True)
        End If

        If locSelectedDeviceItem IsNot Nothing Then
            For Each locItem As ListViewItem In lvwDeviceItems.Items
                Dim locDevItem As IFacessoConversionItem = DirectCast(locItem.Tag, IFacessoConversionItem)
                If locDevItem.AlienElementID = locSelectedDeviceItem.AlienElementID Then
                    locItem.Selected = True
                    locItem.EnsureVisible()
                End If
            Next
        End If
    End Sub

    Protected Overridable Sub btnAdd_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnAdd.Click
        If ucLabourValues.FirstSelectedLabourValue Is Nothing Then
            MessageBox.Show("Bitte wählen Sie zunächst einen REFA-Arbeitswert aus, den Sie an die Geräte-IDs zuweisen möchten!", _
                             "Fehlende Arbeitswertauswahl!", MessageBoxButtons.OK, MessageBoxIcon.Exclamation)
            Return
        End If

        If lvwDeviceItems.SelectedIndices.Count = 0 Then
            MessageBox.Show("Bitte wählen Sie zunächst eine DeviceID aus, der Sie den REFA-Arbeitswert zuweisen möchten!", _
                             "Fehlender Device-ID-Auswahl!", MessageBoxButtons.OK, MessageBoxIcon.Exclamation)
            Return
        End If

        Dim locLabourValue As LabourValueInfo = ucLabourValues.FirstSelectedLabourValue
        Dim locDeviceItem As IFacessoConversionItem = DirectCast(lvwDeviceItems.SelectedItems(0).Tag, IFacessoConversionItem)
        locDeviceItem.HomeElementID = locLabourValue.IDLabourValue
        locDeviceItem.HomeElementName = locLabourValue.LabourValueNumber & ": " & locLabourValue.LabourValueName
        RebuildLists()
    End Sub

    Protected Overridable Sub btnRemove_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnRemove.Click
        If lvwDeviceItems.SelectedIndices.Count = 0 Then
            MessageBox.Show("Bitte wählen Sie zunächst eine DeviceID aus, deren REFA-Arbeitswert-Zuweisung Sie aufheben möchten!", _
                             "Fehlender Device-ID-Auswahl!", MessageBoxButtons.OK, MessageBoxIcon.Exclamation)
            Return
        End If
        Dim locDeviceItem As IFacessoConversionItem = DirectCast(lvwDeviceItems.SelectedItems(0).Tag, IFacessoConversionItem)
        locDeviceItem.HomeElementID = -1
        locDeviceItem.HomeElementName = Nothing
        RebuildLists()
    End Sub

    Protected Overridable Sub btnOK_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnOK.Click
        Me.DialogResult = Windows.Forms.DialogResult.OK
    End Sub

    Protected Overridable Sub btnCancel_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnCancel.Click
        Me.DialogResult = Windows.Forms.DialogResult.Cancel
    End Sub

    Protected Overridable ReadOnly Property BlockDeviceListBuilding() As Boolean
        Get
            Return False
        End Get
    End Property
End Class

File: src/NetFxCSharp/Facesso.Interfaces/frmProductionDataConfigureDialogBase.cs
using Facesso;
using Facesso.Data;
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml.Linq;

namespace Facesso.Interfaces
{
    public partial class frmProductionDataConfigureDialogBase
    {
        private IFacessoImportTaskItem myTaskItem;
        private LabourValueInfoCollection myLabourValues;
        protected bool myAllowMultipleAssignments;
        public DialogResult HandleDialog(IFacessoImportTaskItem TaskItem)
        {
            myTaskItem = TaskItem;
            myLabourValues = LabourValueInfoCollection.GetWorkGroupAssignedLabourValues(FacessoGeneric.LoginInfo.IDSubsidiary, myTaskItem.ForWorkgroup);
            InitializeControls();
            RebuildLists();
            this.ShowDialog();
            return this.DialogResult;
        }

        protected IFacessoImportTaskItem TaskItem
        {
            get
            {
                return myTaskItem;
            }

            set
            {
                myTaskItem = value;
            }
        }

        protected virtual void InitializeControls()
        {
            {
                var __with0 = lvwDeviceItems;
                __with0.Columns.Add("ID", -2, HorizontalAlignment.Left);
                __with0.Columns.Add("Beschreibung", -2, HorizontalAlignment.Left);
                __with0.Columns.Add("Arbeitswert", -2, HorizontalAlignment.Left);
            }

            if (myTaskItem.ConversionItems == null)
            {
                myTaskItem.ConversionItems = myTaskItem.ConversionItemsDelegate.Invoke();
            }

            lblTitel.Text = "Konfiguration für Produktiv-Site:" + System.Environment.NewLine + myTaskItem.ForWorkgroup.ListItemText;
        }

        protected virtual void RebuildLists()
        {
            LabourValueInfo locSelectedLabourValue = null;
            IFacessoConversionItem locSelectedDeviceItem = null;
            try
            {
                locSelectedLabourValue = ucLabourValues.FirstSelectedLabourValue;
                locSelectedDeviceItem = ((IFacessoConversionItem)lvwDeviceItems.SelectedItems[0].Tag);
            }
            catch (Exception ex)
            {
            }

            lvwDeviceItems.BeginUpdate();
            lvwDeviceItems.Items.Clear();
            //Nur neu aufbauen, wenn die ableitende Klasse
            //diese Eigenschaft auf True setzt.
            if (!(BlockDeviceListBuilding))
            {
                if (myTaskItem.ConversionItems != null)
                {
                    foreach (IFacessoConversionItem locItem in myTaskItem.ConversionItems)
                    {
                        ListViewItem locLvwItem = new ListViewItem(locItem.AlienElementID.ToString("000000"));
                        locLvwItem.SubItems.Add(locItem.Itemname);
                        if (locItem.HomeElementID == -1)
                        {
                            locLvwItem.SubItems.Add("- - -");
                        }
                        else
                        {
                            locLvwItem.SubItems.Add(locItem.HomeElementName);
                            locLvwItem.Font = new Font(locLvwItem.Font, FontStyle.Bold);
                        }

                        locLvwItem.Tag = locItem;
                        lvwDeviceItems.Items.Add(locLvwItem);
                    }
                }

                lvwDeviceItems.Columns[0].Width = -2;
                lvwDeviceItems.Columns[1].Width = -2;
                lvwDeviceItems.Columns[2].Width = -2;
            }

            lvwDeviceItems.EndUpdate();
            LabourValueInfoCollection locToAssignList = new LabourValueInfoCollection();
            foreach (LabourValueInfo locItem in myLabourValues)
            {
                locToAssignList.Add(locItem);
            }

            //Elemente, die schon verwendet wurden, nur dann entfernen, wenn
            //Mehrfachzuweisungen nicht erlaubt sind!
            if (!(myAllowMultipleAssignments))
            {
                foreach (ListViewItem locDestItem in lvwDeviceItems.Items)
                {
                    IFacessoConversionItem locItem = ((IFacessoConversionItem)locDestItem.Tag);
                    locToAssignList.Remove(new ActiveDev.IntKey(locItem.HomeElementID));
                }
            }

            ucLabourValues.LabourValues = locToAssignList;
            //Dafür sorgen, dass zuvor Dargestellte wieder angezeigt werden.
            if (locSelectedLabourValue != null)
            {
                ucLabourValues.SelectLabourValue(locSelectedLabourValue, true);
            }

            if (locSelectedDeviceItem != null)
            {
                foreach (ListViewItem locItem in lvwDeviceItems.Items)
                {
                    IFacessoConversionItem locDevItem = ((IFacessoConversionItem)locItem.Tag);
                    if (locDevItem.AlienElementID == locSelectedDeviceItem.AlienElementID)
                    {
                        locItem.Selected = true;
                        locItem.EnsureVisible();
                    }
                }
            }
        }

        protected virtual void btnAdd_Click(System.Object sender, System.EventArgs e)
        {
            if (ucLabourValues.FirstSelectedLabourValue == null)
            {
                MessageBox.Show("Bitte wählen Sie zunächst einen REFA-Arbeitswert aus, den Sie an die Geräte-IDs zuweisen möchten!", "Fehlende Arbeitswertauswahl!", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                return;
            }

            if (lvwDeviceItems.SelectedIndices.Count == 0)
            {
                MessageBox.Show("Bitte wählen Sie zunächst eine DeviceID aus, der Sie den REFA-Arbeitswert zuweisen möchten!", "Fehlender Device-ID-Auswahl!", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                return;
            }

            LabourValueInfo locLabourValue = ucLabourValues.FirstSelectedLabourValue;
            IFacessoConversionItem locDeviceItem = ((IFacessoConversionItem)lvwDeviceItems.SelectedItems[0].Tag);
            locDeviceItem.HomeElementID = locLabourValue.IDLabourValue;
            locDeviceItem.HomeElementName = locLabourValue.LabourValueNumber + ": " + locLabourValue.LabourValueName;
            RebuildLists();
        }

        protected virtual void btnRemove_Click(System.Object sender, System.EventArgs e)
        {
            if (lvwDeviceItems.SelectedIndices.Count == 0)
            {
                MessageBox.Show("Bitte wählen Sie zunächst eine DeviceID aus, deren REFA-Arbeitswert-Zuweisung Sie aufheben möchten!", "Fehlender Device-ID-Auswahl!", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                return;
            }

            IFacessoConversionItem locDeviceItem = ((IFacessoConversionItem)lvwDeviceItems.SelectedItems[0].Tag);
            locDeviceItem.HomeElementID = -1;
            locDeviceItem.HomeElementName = null;
            RebuildLists();
        }

        protected virtual void btnOK_Click(System.Object sender, System.EventArgs e)
        {
            this.DialogResult = System.Windows.Forms.DialogResult.OK;
        }

        protected virtual void btnCancel_Click(System.Object sender, System.EventArgs e)
        {
            this.DialogResult = System.Windows.Forms.DialogResult.Cancel;
        }

        protected virtual bool BlockDeviceListBuilding
        {
            get
            {
                return false;
            }
        }

        public frmProductionDataConfigureDialogBase()
        {
            InitializeComponent();
        }
    }
}